### ISSUE TITLE
Checking branching and memory access based on secret data

### ISSUE THREAD

**Issue by joostrijneveld** [PQClean/PQClean#12]
One of the requirements is that code should not branch based on secret data, and, similarly, not do memory accesses based on secret data.

A first intuition may be to pass an uninitialized secret key to the API functions and check that valgrind does not output any `[..] depends on uninitialised value(s)` errors. This has (at least) two problems:

- the secret key typically contains (part of) the public key as well; labeling all of it as secret is excessive
- rejection sampling based on secret data is not necessarily bad, e.g., sampling from the output of a XOF based on a secret seed

An alternative approach would be to have two `randombytes` functions, i.e., something like `randombytes_public` and `randombytes` (to error on the side of caution, I'd default to the assumed-secret variant). Normally they would both link to the same function, but when running this check the secret variant would not initialize the buffer.

This still leaves open the rejection sampling issue. It seems there is no way around manual inspection in these cases. We could whitelist the lines in which these issues occur, and then filter them out of the valgrind output.

Valgrind has support for suppression files, but the suppression rules seem too generic for our purposes. For example, consider [the rejection sampling in Kyber](https://github.com/PQClean/PQClean/blob/4cac434f47f5e2c99839cf8ea4f3a4241f5eb97c/crypto_kem/kyber768/clean/indcpa.c#L162), which triggers the following valgrind error and corresponding suppression rule

```
==6102== Conditional jump or move depends on uninitialised value(s)
==6102==    at 0x11439B: PQCLEAN_KYBER768_gen_matrix (indcpa.c:162)
==6102==    by 0x114769: PQCLEAN_KYBER768_indcpa_enc (indcpa.c:254)
==6102==    by 0x114D2D: PQCLEAN_KYBER768_crypto_kem_dec (kem.c:110)
==6102==    by 0x116EAC: test_null_sk (null_secretkey.c:32)
==6102==    by 0x116F10: main (null_secretkey.c:44)
==6102== 
{
   <insert_a_suppression_name_here>
   Memcheck:Cond
   fun:PQCLEAN_KYBER768_gen_matrix
   fun:PQCLEAN_KYBER768_indcpa_enc
   fun:PQCLEAN_KYBER768_crypto_kem_dec
   fun:test_null_sk
   fun:main
}
```

We could have a file with all whitelisted files and lines, but that quickly becomes a hassle to keep up to date when the line numbers change.

We could also include inline suppression markers (similar to `#noqa`), and then use this to filter valgrind output. Such markers are perhaps easily overlooked when doing code review, so we would probably want to search for these automatically and highlight them as part of the CI output (through a webhook and a comment on the PR?).

@mkannwischer and I briefly discussed this, but figured it would be worthwhile to hear some more thoughts on this before implementing anything concrete.

**Comment by cryptojedi** [comment]
Joost Rijneveld <notifications@github.com> wrote:
> One of the requirements is that code should not branch based on secret
> data, and, similarly, not do memory accesses based on secret data.
> 
> A first intuition may be to pass an uninitialized secret key to the
> API functions and check that valgrind does not output any `[..]
> depends on uninitialised value(s)` errors. This has (at least) two
> problems:
> 
> - the secret key typically contains (part of) the public key as well;
> labeling all of it as secret is excessive - rejection sampling based
> on secret data is not necessarily bad, e.g., sampling from the output
> of a XOF based on a secret seed
> 
> An alternative approach would be to have two `randombytes` functions,
> i.e., something like `randombytes_public` and `randombytes` (to error
> on the side of caution, I'd default to the assumed-secret variant).
> Normally they would both link to the same function, but when running
> this check the secret variant would not initialize the buffer.
> 
> This still leaves open the rejection sampling issue. It seems there is
> no way around manual inspection in these cases. We could whitelist the
> lines in which these issues occur, and then filter them out of the
> valgrind output.

These valgrind-based tests should certainly only generate warnings, so
that if errors are reported, you need to inspect manually. However, for
schemes that do pass those tests, the effort is much smaller than with
full manual inspection. We might want to require comments in the code at
the respective spots explaining why valgrind with uninitialized secret
data complains.

I'm not sure about the randombytes_public function. That would slightly
complicate integration into other code bases, right?

**Comment by thomwiggers** [comment]
Perhaps this is just too complicated to do completely automatically as part of CI. The script(s) that run these checks could perhaps just be run before merging a PR as part of the 'check the PR' workflow (which we also need to establish). The valgrind issues that pop up then are then easily determined to be valid or not.

**Comment by joostrijneveld** [comment]
Good points. I agree that this is going to result in manual effort anyway, and having `randombytes_public` then indeed only complicates things for users without gaining us much.

What I was hoping to achieve with whitelisting markers is that it could be an error in the CI process if there are valgrind errors on places where markers are not present, essentially formalizing requiring a comment at the respective spots in the code (and, informally, requiring a bit more explanation than a simple `PQCLEAN_SECRET_ACCESS_OK`). That may indeed not be worth the extra complexity either.

**Comment by dstebila** [comment]
@cryptojedi working with colleagues to get a better version of ctverif that works more automatically and has fewer false positives/negatives.

**Comment by gmaxwell** [comment]
In libsecp256k1 we've used valgrind for constant time testing for years and recently made it part of CI (after I got tired of other contributors being unable to remember that we had this tool and should be using it. :P).  I have yet to encounter a single false positive in our configuration other than just valgrind not working at all (we have, however, been struggling a littl with compilers making code actually non-constant time).

To accomplish this, we make use of the valgrind memcheck instrumentation in a test harness to _explicitly_ mark secret memory:

https://github.com/bitcoin-core/secp256k1/blob/37dba329c6cb0f7a4228a11dc26aa3a342a3a5d0/src/valgrind_ctime_test.c#L47

This avoids any problems with having a test with undefined behaviour the compiler might decide to get clever with.

We also use the instrumentation to construct a declassification function to mark data that through cryptographic magic has been rendered no longer secret:   https://github.com/bitcoin-core/secp256k1/blob/37dba329c6cb0f7a4228a11dc26aa3a342a3a5d0/src/secp256k1.c#L225

I also made some custom patches to valgrind to expand the set of detected operations, but they're not really relevant for us-- they might be relevant to you-- in particular, stock memcheck doesn't protest about tainted memory going into a division operator. 

I just stumbled across your repository randomly while googling to look for reasons why GCC 10+PPC64LE+ Os optimization = valgrind puking all over everything.

**Comment by dstebila** [comment]
Hi @gmaxwell, I saw your post here and tried applying your technique to FrodoKEM which is included in PQClean (although I applied it in a slightly different version for now).  See https://github.com/microsoft/PQCrypto-LWEKE/compare/master...dstebila:ds-valgrind?expand=1.  Does that seem like the right application of your technique?

**Comment by gmaxwell** [comment]
@dstebila  (sorry for the reply delay)  That looks correct, is it working usefully for you?

**Comment by dstebila** [comment]
> @dstebila (sorry for the reply delay) That looks correct, is it working usefully for you?

Yup, it successfully identified something that I knew had a secret dependent branch and our proposed fix seemed to pass it.

> One thing to watch out for is that the -O0 may actually cause the compiler to emit branches where it wouldn't otherwise or vice versa.

Right... but as I understand it, all optimization levels may or may not generate some code.  So I'd have to run Valgrind for all the optimization levels I envision the code being used with in order to have assurances at each of those levels?

**Comment by gmaxwell** [comment]
And on all compilers and all target architectures that you expect to have assurances of.  :(

On the plus side, if it's an outright bug in your code it's more likely to trigger a failure in many configurations vs if its a case of the compiler being overly clever which might just trigger on one platform.

Given that you can't test every conceivable configuration (especially not future compilers that don't exist yet) if you test the most likely setups you're at least testing something that matters to some users.

I'm glad that it's also working for you!

**Comment by thomwiggers** [comment]
The KPQClean project integrated Valgrind-based constant-time checking: https://github.com/kpqc-cryptocraft/KPQClean/tree/main/Valgrind_report_constant_time

**Comment by mkannwischer** [comment]
We have a patch to Valgrind that can also allows checking for variable-latency instructions like divisions in case someone is interested: https://kyberslash.cr.yp.to/papers.html. 
We are trying the upstream that to Valgrind, but we have not heard back yet from the Valgrind developers. 

I'm not very eager to spent many cycles on PQClean these days though.

**Comment by thomwiggers** [comment]
> I'm not very eager to spent many cycles on PQClean these days though.

Same, I was mostly making a note of it

**Comment by thomwiggers** [comment]
Please note that we are currently planning to archive the PQClean repository. The current schedule has us marking PQClean as read-only in July 2026. Please refer to issue #604 for more information and if you have any feedback.
